### PR TITLE
Enable API Gateway logging in CloudWatch (#3054)

### PR DESCRIPTION
**Pull request by achave11**
https://github.com/DataBiosphere/azul/issues/3054

Author

- [x] PR title references issue
- [x] PR title matches issue title (preceded by `Fix: ` for bugs)   <sub>or there is a good reason why they're different</sub>
- [x] Title of main commit references issue
- [x] PR is connected to Zenhub issue and description links to issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>
- [x] [Added announcement to PR description](https://github.com/DataBiosphere/azul/pull/3415#issuecomment-916447381)                  <sub>or this PR does not require announcement</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passed in sandbox                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [ ] Made announcement requested by author                 <sub>or PR description does not contain an announcement</sub>
- [ ] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] [Follow README.md instructions in this PR for `prod` deployment](https://github.com/DataBiosphere/azul/pull/3415#issuecomment-916447381)
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

**Comment by achave11** [comment]
Note that this PR contains additions to the `README.md`. These additions contain instructions to be followed for the `prod` deployment, `dev` was addressed in conjunction with this changes.

File: README.md [hunk 1]
```diff
@@ -414,6 +414,34 @@ temporary credentials were issued in. To account for the region specificity of
...
+
+8. For *Role name* enter `azul-api_gateway`, ensure *Policies* are as prescribed
+in step 6 and click *Create role*
+
+9. Be sure to copy the *Role ARN* for the newly created `azul-api_gateway` role
+and go to the *API Gateway* console
+
+10. Select any available Gateway resource e.g. `azul-{lambda}-{stage}`, click on
+*Settings* at the bottom of the left menu, paste the copied *Role ARN* into the
+*CloudWatch log role ARN* and click *Save*
```
Comment (jessebrennan) [inline README.md:443]: I'm confused about this step. You're supposed to set the ARN for *any* available Gateway resource? You do it for one and it activates the logging role for everything? Can you explain how this works?

File: terraform/api_gateway.tf.json.template.py [hunk 2]
```diff
@@ -157,28 +169,23 @@ def for_domain(cls, domain):
...
+                            "command": ' '.join(map(shlex.quote, [
+                                "python",
+                                config.project_root + "/scripts/log_api_gateway.py",
+                                "${module.chalice_%s.RestAPIId}" % lambda_.name,
+                                config.deployment_stage,
+                                "${aws_cloudwatch_log_group.%s.arn}" % lambda_.name
+                            ]))
                         }
                     }
-                } if config.enable_monitoring else {
```
Comment (jessebrennan) [inline terraform/api_gateway.tf.json.template.py:179]: Why this change? Did you test deploying this with AZUL_ENABLE_MONITORING turned off?

**Review by jessebrennan — COMMENTED** [review]
Mostly looks good, I just had a couple questions.

File: README.md (hunk 1 above)
Comment (achave11) [reply README.md:443]: note to self: Add `to modify region wide settings do…` or something like it.

File: terraform/api_gateway.tf.json.template.py (hunk 2 above)
Comment (achave11) [reply terraform/api_gateway.tf.json.template.py:179]: CloudWatch logs should be enabled at all times. Additionally, enabling `AZUL_ENABLE_MONITORING` sets up other monitoring resources (health checks, alarms, etc.) which was beyond the scope of this ticket.

**Review by jessebrennan — APPROVED** [review]

File: README.md [hunk 3]
```diff
@@ -422,6 +422,36 @@ the *Permanently delete previous versions of objects* checkbox, and enter *30*
 for *Number of days after objects become previous versions*. Then click *Create
 rule.*
 
+To enable CloudWatch logs for API Gateway, an IAM role must be created. This
+needs to be done once per region. Deployments with CloudWatch enabled  are
```
Comment (hannes-ucsc) [inline README.md:426]:
```suggestion
needs to be done once per region. Deployments with CloudWatch enabled are
```

File: README.md [hunk 4]
```diff
@@ -422,6 +422,36 @@ the *Permanently delete previous versions of objects* checkbox, and enter *30*
...
+automatically covered by this configuration.
```
Comment (hannes-ucsc) [inline README.md:427]: Don't understand the last sentence.

File: README.md [hunk 5]
```diff
@@ -422,6 +422,36 @@ the *Permanently delete previous versions of objects* checkbox, and enter *30*
...
+
+6. Ensure *AmazonAPIGatewayPushToCloudWatchLogs* is listed exclusively as the
+policy and click *Next: Tags*
+
+7. Add *name: azul-api_gateway* as a tag and click *Next: Review*
+
+8. For *Role name* enter `azul-api_gateway`, ensure *Policies* are as prescribed
+in step 6 and click *Create role*
+
+9. Be sure to copy the *Role ARN* for the newly created `azul-api_gateway` role
```
Comment (hannes-ucsc) [inline README.md:447]:
```suggestion
9. Copy the *Role ARN* for the newly created `azul-api_gateway` role
```

File: README.md [hunk 6]
```diff
@@ -422,6 +422,36 @@ the *Permanently delete previous versions of objects* checkbox, and enter *30*
...
+and go to the *API Gateway* console
+
+10. Select any available Gateway resource (it doesn't matter which because this
```
Comment (hannes-ucsc) [inline README.md:450]: What if the current region has no gateways? Maybe there needs to be an intro paragraph stating that this must be done after at least one deployment was created in the configured AWS account and region.

Comment (hannes-ucsc) [reply README.md:450]:
```suggestion
10. Select any available API Gateway resource (it doesn't matter which because this 
```

File: terraform/api_gateway.tf.json.template.py [hunk 7]
```diff
@@ -105,6 +105,18 @@ def for_domain(cls, domain):
                     "certificate_arn": "${aws_acm_certificate_validation.%s_%i.certificate_arn}" % (lambda_.name, i)
                 } for i, domain in enumerate(lambda_.domains)
             },
+            "aws_api_gateway_method_settings": {
+                f"{lambda_.name}_{i}": {
+                    "rest_api_id": "${module.chalice_%s.RestAPIId}" % lambda_.name,
+                    "stage_name": "${module.chalice_%s.stage_name}" % lambda_.name,
+                    "method_path": "*/*",
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:112]: Why not just `*`? Is the pattern behavior documented somewhere?

File: terraform/api_gateway.tf.json.template.py [hunk 8]
```diff
@@ -105,6 +105,18 @@ def for_domain(cls, domain):
...
+                    "settings": {
+                        "metrics_enabled": True,
+                        "data_trace_enabled": True if config.debug == 2 else False,
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:115]:
```suggestion
                        "data_trace_enabled": config.debug == 2,
```

File: terraform/api_gateway.tf.json.template.py [hunk 9]
```diff
@@ -105,6 +105,18 @@ def for_domain(cls, domain):
...
+                        "logging_level": "INFO" if config.debug != 0 else "ERROR"
```
Comment (hannes-ucsc) [inline terraform/api_gateway.tf.json.template.py:116]:
```suggestion
                        "logging_level": "ERROR" if config.debug == 0 else "INFO"
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: terraform/api_gateway.tf.json.template.py (hunk 7 above)
Comment (achave11) [reply terraform/api_gateway.tf.json.template.py:112]: [Terraform `method_path`](https://registry.terraform.io/providers/hashicorp/aws/latest/docs/resources/api_gateway_method_settings#method_path) documents the usage of `*/*`.

File: README.md [hunk 10]
```diff
@@ -422,6 +422,38 @@ the *Permanently delete previous versions of objects* checkbox, and enter *30*
 for *Number of days after objects become previous versions*. Then click *Create
 rule.*
 
+To enable CloudWatch logs for API Gateway, an IAM role must be created. This
+can be done after at least one deployment is created in the configured AWS
+account and region, and is required once per region. With this configuration all
+deployments in API Gateway assume the created IAM role, which allows API Gateway
+to write logs in CloudWatch.
```
Comment (hannes-ucsc) [inline README.md:429]:
```suggestion
To enable CloudWatch logs for API Gateway, an IAM role must be created and configured in the API Gateway console. This
must be done for each AWS account and region, after at least one Azul deployment has been created in that account and region. Once these steps have been completed, all
API Gateway instances in a region assume the resulting IAM role, allowing those instances
to write to CloudWatch Logs in that region.
```

File: terraform/api_gateway.tf.json.template.py (hunk 7 above)
Comment (hannes-ucsc) [reply terraform/api_gateway.tf.json.template.py:112]:
```suggestion
                    "method_path": "*/*",  # every URL path, every HTTP method
```

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

**Review by hannes-ucsc — APPROVED** [review]
